concent_api/core/payments/storage.py

```diff
@@ -0,0 +1,145 @@
+                f'Current={global_transaction_state.nonce}, '
+                f'Transaction={tx.nonce}'
+            )
+
+        pending_ethereum_transaction = PendingEthereumTransaction(
+            nonce       = tx.nonce,
+            gasprice    = tx.gasprice,
+            startgas    = tx.startgas,
+            value       = tx.value,
+            v           = tx.v,
+            r           = tx.r,
+            s           = tx.s,
+            data        = tx.data,
+            to          = tx.to,
+        )
+        pending_ethereum_transaction.full_clean()
+        pending_ethereum_transaction.save()
+
+        global_transaction_state.nonce += 1
+        global_transaction_state.full_clean()
+        global_transaction_state.save()
+
+    @transaction.atomic(using='control')
+    def remove_tx(self, nonce: int) -> None:
+        """
+        Remove the transaction after it's been confirmed and doesn't have
+        to be tracked anymore.
+        """
+
+        # This is called here only to lock GlobalTransactionState
+        self._get_locked_global_transaction_state()
+
+        try:
+            pending_ethereum_transaction = PendingEthereumTransaction.objects.get(nonce=nonce)
+            pending_ethereum_transaction.delete()
+            logger.info(f'Successfully removed PendingEthereumTransaction with nonce {nonce}.')
+        except PendingEthereumTransaction.DoesNotExist:
+            logger.error(f'Trying to remove PendingEthereumTransaction with nonce {nonce} but it does not exist.')
+            raise
+
+    @transaction.atomic(using='control')
+    def revert_last_tx(self) -> None:
+        """
+        Remove the last transaction that was added.
+        This shouldn't be ever called if everything is being used correctly,
+        i.e. we don't try to send invalid transactions.
+        """
+        global_transaction_state = self._get_locked_global_transaction_state()
+
+        try:
+            pending_ethereum_transaction = PendingEthereumTransaction.objects.get(
+                nonce=global_transaction_state.nonce - 1
+            )
+            pending_ethereum_transaction.delete()
+            logger.info(
+                f'Successfully reverted last PendingEthereumTransaction with nonce {pending_ethereum_transaction.nonce}.'
+            )
+        except GlobalTransactionState.DoesNotExist:
+            logger.error(
+                f'Trying to revert last PendingEthereumTransaction with nonce {global_transaction_state.nonce - 1} but it does not exist.'
+            )
+            raise
+
+    @staticmethod
+    def _get_locked_global_transaction_state() -> GlobalTransactionState:
+        try:
+            return GlobalTransactionState.objects.select_for_update().get(
+                pk=0,
+            )
+        except GlobalTransactionState.DoesNotExist:
+            logger.error(f'Trying to get GlobalTransactionState but it does not exist.')
+            raise
```

Comment:
I think we will need also method which will get oldest transaction which hasn't been proceed yet. @cameel what do you think?

Reply:
The `nonce` does that. By checking its value you can figure out which transaction was added last.

Reply:
@kbeker Also, we are not implementing our custom interface for this, but we follow https://github.com/golemfactory/golem-smart-contracts-interface/blob/v1.4.2/golem_sci/transactionsstorage.py#L16-L53